common/src/View/SwapNodeContentsCommand.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ Copyright (C) 2010-2017 Kristian Duske
+
+ This file is part of TrenchBroom.
+
+ TrenchBroom is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ TrenchBroom is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with TrenchBroom. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "SwapNodeContentsCommand.h"
+
+#include "Model/Node.h"
+#include "View/MapDocumentCommandFacade.h"
+
+#include <kdl/vector_utils.h>
+
+namespace TrenchBroom {
+    namespace View {
+        const Command::CommandType SwapNodeContentsCommand::Type = Command::freeType();
+
+        SwapNodeContentsCommand::SwapNodeContentsCommand(const std::string& name, std::vector<std::tuple<Model::Node*, std::unique_ptr<Model::Node>>> nodes) :
+        DocumentCommand(Type, name),
+        m_nodes(std::move(nodes)) {}
+
+        SwapNodeContentsCommand::~SwapNodeContentsCommand() = default;
+
+        std::unique_ptr<CommandResult> SwapNodeContentsCommand::doPerformDo(MapDocumentCommandFacade* document) {
+            document->performSwapNodeContents(m_nodes);
+            return std::make_unique<CommandResult>(true);
+        }
+
+        std::unique_ptr<CommandResult> SwapNodeContentsCommand::doPerformUndo(MapDocumentCommandFacade* document) {
+            document->performSwapNodeContents(m_nodes);
+            return std::make_unique<CommandResult>(true);
+        }
+
+        bool SwapNodeContentsCommand::doCollateWith(UndoableCommand* command) {
+            auto* other = static_cast<SwapNodeContentsCommand*>(command);
+
+            auto myNodes = kdl::vec_transform(m_nodes, [](const auto& t) { return std::get<0>(t); });
+            auto theirNodes = kdl::vec_transform(other->m_nodes, [](const auto& t) { return std::get<0>(t); });
+
+            kdl::vec_sort(myNodes);
+            kdl::vec_sort(theirNodes);
+            return myNodes == theirNodes;
```

Comment:
This doesn't look like it's collating the commands.

I don't think this is correct "can collate" check either, shouldn't it always be possible to collate?

AFAIK the correct collation implementation would look like: if you view m_nodes as a map, "add all of the key/values from the other command's map to our map".

Reply:
This works because the commands are in a "done" state when this is called. So the current state of the document already contains the changes from the given command at this point. So at this point:

- `this` command contains the state of the nodes before this command was executed, which is the state we want to go to on undo.
- The given command contains the state of the nodes after this command was executed, which is invisible when undoing / redoing a transaction.
- The document contains the state of the nodes after the given command was executed, which is the state we want to go to on redo (after `this` was undone).

Since we only collate if both commands affect the same nodes, we can just forget about the second command, as it only contains some immediate state we don't need anymore.

We could possibly relax the condition for collation, but I don't think we need to. We want to collate series of commands that affect the same nodes. Do you see a situation where we want to collate commands where the second command affects different sets of nodes?

Reply:
Ah right, I forgot how the collation api worked.

It still looks wrong though: in the case of an undo followed by a redo, you'll lose the changes that were in the collated command. m_nodes needs to be updated in order for that to work. e.g. other implementations like `MoveBrushFacesCommand::doCollateWith` always have to update `this` command's delta to include the delta from the `other` command, so that the "undo, then redo" case works.

Reply:
> We could possibly relax the condition for collation, but I don't think we need to. We want to collate series of commands that affect the same nodes. 
> Do you see a situation where we want to collate commands where the second command affects different sets of nodes?

Some cases where it's not the same set of nodes but would useful to collate would be selection painting, or texture painting across brushes with Alt+drag.

This isn't necessary (collation is just a undo/redo performance optimization + memory use optimization, right?) but might be nice to have. We don't currently collate SelectionCommand.

Reply:
> It still looks wrong though: in the case of an undo followed by a redo, you'll lose the changes that were in the collated command. m_nodes needs to be updated in order for that to work. e.g. other implementations like MoveBrushFacesCommand::doCollateWith always have to update this command's delta to include the delta from the other command, so that the "undo, then redo" case works.

No, I don't think that's true. When the collated command is undone, the nodes in the document are swapped with the nodes in the command, so the command will then contain the state the document was in. Then when you redo, you swap that state back into the document.

Reply:
> Some cases where it's not the same set of nodes but would useful to collate would be selection painting, or texture painting across brushes with Alt+drag.

Right, we could explore if we collate in such cases, too.

Reply:
> When the collated command is undone, the nodes in the document are swapped with the nodes in the command, so the command will then contain the state the document was in. Then when you redo, you swap that state back into the document.

Ah thanks I had missed that.

Reply:
Yeah it took me a few attempts to get this right, it's hard to reason about because the commands are already done.